bmq-sdk/src/main/java/com/bloomberg/bmq/Subscription.java

```diff
@@ -0,0 +1,311 @@
+     */
+    public static Subscription createDefault() {
+        return new Subscription();
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("[ maxUnconfirmedMessages: ")
+                .append(getMaxUnconfirmedMessages())
+                .append(", maxUnconfirmedBytes: ")
+                .append(getMaxUnconfirmedBytes())
+                .append(", consumerPriority: ")
+                .append(getConsumerPriority())
+                .append(" ]");
+        return sb.toString();
+    }
+
+    /** A helper class to create immutable {@code Subscription} with custom settings. */
+    public static class Builder {
+        private Optional<Long> maxUnconfirmedMessages;
+        private Optional<Long> maxUnconfirmedBytes;
+        private Optional<Integer> consumerPriority;
+        private SubscriptionExpression.Builder expressionBuilder;
+
+        /**
+         * Returns subscription parameters based on this object properties.
+         *
+         * @return Subscription immutable subscription parameters object
+         */
+        public Subscription build() {
+            return new Subscription(this);
+        }
+
+        /**
+         * Sets maximum number of outstanding messages that can be sent by the broker without being
+         * confirmed for this subscription.
+         *
+         * @param maxUnconfirmedMessages maximum number of outstanding messages
+         * @return Builder this object
+         */
+        public Subscription.Builder setMaxUnconfirmedMessages(long maxUnconfirmedMessages) {
+            this.maxUnconfirmedMessages = Optional.of(maxUnconfirmedMessages);
+            return this;
+        }
+
+        /**
+         * Sets maximum accumulated bytes of all outstanding messages that can be sent by the broker
+         * without being confirmed for this subscription.
+         *
+         * @param maxUnconfirmedBytes maximum accumulated bytes of all outstanding messages without
+         *     being confirmed
+         * @return Builder this object
+         */
+        public Subscription.Builder setMaxUnconfirmedBytes(long maxUnconfirmedBytes) {
+            this.maxUnconfirmedBytes = Optional.of(maxUnconfirmedBytes);
+            return this;
+        }
+
+        /**
+         * Sets priority of a consumer for this subscription with respect to delivery of messages.
+         *
+         * @param consumerPriority priority of a consumer
+         * @return Builder this object
+         */
+        public Subscription.Builder setConsumerPriority(int consumerPriority) {
+            this.consumerPriority = Optional.of(consumerPriority);
+            return this;
+        }
+
+        /**
+         * Sets expression text for this subscription.
+         *
+         * @param expression expression text
+         * @return Builder this object
+         */
+        public Subscription.Builder setExpressionText(String expression) {
```

Comment:
`setExpression` here will be inconvenient here, because it makes API usage more grotesque.

Compare:
```
Subscription s3 =
        Subscription.builder()
                .setExpressionText("firmId == \"BAR\" && price >= 25")
                .setUserData("bar_high_price")
                .build();
```

With:
```
Subscription s3 =
        Subscription.builder()
                .setExpression(new SubscriptionExpression("firmId == \"BAR\" && price >= 25"))
                .setUserData("bar_high_price")
                .build();
```

Or worse:
```
Subscription s3 =
        Subscription.builder()
                .setExpression(
                        SubscriptionExpression.builder()
                                .setText("firmId == \"BAR\" && price >= 25")
                                .setVersion(Version.e_VERSION_1)
                                .build())
                .setUserData("bar_high_price")
                .build();
```

From my perspective, we don't expect a lot of duplicating setters here, just text and version. But it flattens `Builder` usage and makes it more convenient for an end-user.